File: Frontend/src/app/Services/MyLearning/my-learning.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { map } from 'rxjs/operators';
import { environment } from '../../../environments/environment';
import { ApiResponse } from '../../Dtos/ApiResponse';
import {
  MyLearningResponseDto,
  EnrolledCourseDto,
  MyLearningStatsDto
} from '../../Dtos/MyLearningDtos';

@Injectable({
  providedIn: 'root'
})
export class MyLearningService {
  private api = `${environment.apiUrl}`;

  constructor(private http: HttpClient) {}

  /**
   * Get complete learning dashboard for a user
   */
  getMyLearningDashboard(userId: number): Observable<MyLearningResponseDto> {
    return this.http.get<ApiResponse<MyLearningResponseDto>>(
      `${this.api}/api/my-learning/dashboard/${userId}`
    ).pipe(
      map(response => response.data)
    );
  }

  /**
   * Get all enrolled courses for a user
   */
  getEnrolledCourses(userId: number): Observable<EnrolledCourseDto[]> {
    return this.http.get<ApiResponse<EnrolledCourseDto[]>>(
      `${this.api}/api/my-learning/courses/${userId}`
    ).pipe(
      map(response => response.data)
    );
  }

  /**
   * Get learning statistics for a user
   */
  getLearningStats(userId: number): Observable<MyLearningStatsDto> {
    return this.http.get<ApiResponse<MyLearningStatsDto>>(
      `${this.api}/api/my-learning/stats/${userId}`
    ).pipe(
      map(response => response.data)
    );
  }

  /**
   * Get recently accessed courses (Continue Learning)
   */
  getContinueLearning(userId: number, limit: number = 5): Observable<EnrolledCourseDto[]> {
    return this.http.get<ApiResponse<EnrolledCourseDto[]>>(
      `${this.api}/api/my-learning/continue/${userId}?limit=${limit}`
    ).pipe(
      map(response => response.data)
    );
  }

  /**
   * Get courses with expiring access
   */
  getUpcomingDeadlines(userId: number, daysThreshold: number = 30): Observable<EnrolledCourseDto[]> {
    return this.http.get<ApiResponse<EnrolledCourseDto[]>>(
      `${this.api}/api/my-learning/deadlines/${userId}?daysThreshold=${daysThreshold}`
    ).pipe(
      map(response => response.data)
    );
  }

  /**
   * Get enrolled course details with progress
   */
  getEnrolledCourseDetails(userId: number, courseId: number): Observable<EnrolledCourseDto> {
    return this.http.get<ApiResponse<EnrolledCourseDto>>(
      `${this.api}/api/my-learning/course/${userId}/${courseId}`
    ).pipe(
      map(response => response.data)
    );
  }
}

File: Backend/src/main/java/com/example/e_learning_system/Service/Interfaces/MyLearningService.java
package com.example.e_learning_system.Service.Interfaces;

import com.example.e_learning_system.Dto.MyLearningDtos.EnrolledCourseDto;
import com.example.e_learning_system.Dto.MyLearningDtos.MyLearningResponseDto;
import com.example.e_learning_system.Dto.MyLearningDtos.MyLearningStatsDto;

import java.util.List;

/**
 * Service interface for My Learning functionality
 */
public interface MyLearningService {
    
    /**
     * Get complete learning dashboard for a user
     * @param userId User ID
     * @return MyLearningResponseDto with stats and enrolled courses
     */
    MyLearningResponseDto getMyLearningDashboard(Integer userId);
    
    /**
     * Get all enrolled courses for a user with progress
     * @param userId User ID
     * @return List of enrolled courses with progress
     */
    List<EnrolledCourseDto> getEnrolledCourses(Integer userId);
    
    /**
     * Get learning statistics for a user
     * @param userId User ID
     * @return Learning statistics
     */
    MyLearningStatsDto getLearningStats(Integer userId);
    
    /**
     * Get recently accessed courses (Continue Learning)
     * @param userId User ID
     * @param limit Number of courses to return
     * @return List of recently accessed courses
     */
    List<EnrolledCourseDto> getContinueLearning(Integer userId, int limit);
    
    /**
     * Get courses with expiring access
     * @param userId User ID
     * @param daysThreshold Number of days to check for expiration
     * @return List of courses with expiring access
     */
    List<EnrolledCourseDto> getUpcomingDeadlines(Integer userId, int daysThreshold);
    
    /**
     * Get enrolled course details with progress
     * @param userId User ID
     * @param courseId Course ID
     * @return Enrolled course with progress
     */
    EnrolledCourseDto getEnrolledCourseDetails(Integer userId, Integer courseId);
}
